src/NHibernate/Transaction/AdoTransaction.cs

```diff
@@ -190,20 +190,19 @@ public void Commit()
 
 				log.Debug("Start Commit");
 
+				//Hibernate calls flush from inside BeforeTransactionCompletion
+				if (session.FlushMode != FlushMode.Manual)
+					session.Flush();
+				session.BeforeTransactionCompletion(this);
```

Comment:
Because of distributed transactions. There is a slightly different approach (`ShouldFlushFromDistributedTransaction` + `FlushingFromDtcTransaction`.